File: src/tachikoma/workflows/composition.py
"""Pure composition helpers and in-memory dataclasses for workflow composition.

No SDK / DB / async dependencies. Keeps validation testable in isolation.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from loguru import logger

from tachikoma.workflows.definition import StepDefinition, WorkflowDefinition

_log = logger.bind(component="workflows")


# ---------------------------------------------------------------------------
# Resolution helper
# ---------------------------------------------------------------------------

_VERTEX = tuple[str, str]  # (skill_name, workflow_name)


def resolve_composes(composes_str: str, parent_skill_name: str) -> tuple[str, str]:
    """Resolve a ``composes`` frontmatter value to ``(skill_name, workflow_name)``.

    Same-skill: ``"process-inbox-note"`` → ``(parent_skill_name, "process-inbox-note")``
    Cross-skill: ``"review-skill/process-inbox-note"`` → ``("review-skill", "process-inbox-note")``

    Raises ``ValueError`` on empty or malformed input.
    """
    if not composes_str or not composes_str.strip():
        raise ValueError("composes value must be a non-empty string")

    composes_str = composes_str.strip()

    if "/" in composes_str:
        parts = composes_str.split("/", 1)
        skill, workflow = parts[0], parts[1]
        if not skill or not workflow:
            raise ValueError(f"Malformed composes value: '{composes_str}'")
        # Reject values with additional slashes inside the workflow name
        if "/" in workflow:
            raise ValueError(f"Malformed composes value (multiple slashes): '{composes_str}'")
        return skill, workflow

    return parent_skill_name, composes_str


def _composition_edge(step: StepDefinition) -> str | None:
    """Yield the composition edge value for a step (composes or loop, at most one).

    Mutex enforcement runs upstream in the registry pre-pass, guaranteeing at
    most one of the two is set for any surviving step.
    """
    if step.composes:
        return step.composes
    if step.loop:
        return step.loop
    return None


# ---------------------------------------------------------------------------
# Cycle detection (three-color DFS)
# ---------------------------------------------------------------------------

def detect_cycles(
    workflows: dict[_VERTEX, WorkflowDefinition],
) -> list[list[_VERTEX]]:
    """Return a list of strongly-connected components (each a list of vertices).

    A single self-loop is one SCC of size 1; a mutual cycle A↔B is one SCC
    of size 2.  Non-cyclic workflows are not returned.

    Builds the adjacency map by iterating each workflow's steps and resolving
    ``composes`` fields.  Steps whose ``composes`` value is malformed are
    skipped (caught later by reference validation).
    """
    WHITE, GRAY, BLACK = 0, 1, 2  # noqa: N806

    # Build adjacency map
    adj: dict[_VERTEX, list[_VERTEX]] = {v: [] for v in workflows}

    for (skill, wf_name), wf_def in workflows.items():
        for step in wf_def.steps:
            edge = _composition_edge(step)
            if not edge:
                continue
            try:
                target_skill, target_wf = resolve_composes(edge, skill)
            except ValueError:
                continue
            target = (target_skill, target_wf)
            # Only add edges to vertices in the graph (self-loops are valid)
            if target in workflows:
                adj[(skill, wf_name)].append(target)

    color: dict[_VERTEX, int] = dict.fromkeys(workflows, WHITE)
    stack: list[_VERTEX] = []
    sccs: list[list[_VERTEX]] = []

    def dfs(v: _VERTEX) -> None:
        color[v] = GRAY
        stack.append(v)
        for neighbour in adj.get(v, []):
            if color[neighbour] == GRAY:
                # Found a back-edge — extract the cycle
                idx = stack.index(neighbour)
                sccs.append(list(stack[idx:]))
            elif color[neighbour] == WHITE:
                dfs(neighbour)
        stack.pop()
        color[v] = BLACK

    for v in workflows:
        if color[v] == WHITE:
            dfs(v)

    return sccs


# ---------------------------------------------------------------------------
# Reference validation (fixed-point iteration)
# ---------------------------------------------------------------------------

def validate_references(
    workflows: dict[_VERTEX, WorkflowDefinition],
    already_rejected: set[_VERTEX],
) -> set[_VERTEX]:
    """Return the set of newly-rejected vertices (callers union with ``already_rejected``).

    A workflow is rejected when one of its composition steps references a
    target that is missing, already rejected, or has zero steps.  Cascading
    rejection converges via fixed-point iteration.
    """
    rejected = set(already_rejected)

    changed = True
    while changed:
        changed = False
        for vertex, wf_def in list(workflows.items()):
            if vertex in rejected:
                continue
            for step in wf_def.steps:
                edge = _composition_edge(step)
                if not edge:
                    continue
                edge_kind = "loop" if step.loop else "composes"
                try:
                    target_skill, target_wf = resolve_composes(
                        edge, vertex[0]
                    )
                except ValueError:
                    _log.warning(
                        "Workflow rejected: malformed {kind} value: "
                        "skill={skill}, workflow={workflow}, step={step}, value={value}",
                        kind=edge_kind,
                        skill=vertex[0],
                        workflow=vertex[1],
                        step=step.id,
                        value=edge,
                    )
                    rejected.add(vertex)
                    changed = True
                    break

                target = (target_skill, target_wf)

                if target not in workflows:
                    _log.warning(
                        "Workflow rejected: {kind} target not found: "
                        "skill={skill}, workflow={workflow}, step={step}, "
                        "target_skill={ts}, target_workflow={tw}",
                        kind=edge_kind,
                        skill=vertex[0],
                        workflow=vertex[1],
                        step=step.id,
                        ts=target_skill,
                        tw=target_wf,
                    )
                    rejected.add(vertex)
                    changed = True
                    break

                if target in rejected:
                    _log.warning(
                        "Workflow rejected: {kind} target was itself rejected: "
                        "skill={skill}, workflow={workflow}, step={step}, "
                        "target_skill={ts}, target_workflow={tw}",
                        kind=edge_kind,
                        skill=vertex[0],
                        workflow=vertex[1],
                        step=step.id,
                        ts=target_skill,
                        tw=target_wf,
                    )
                    rejected.add(vertex)
                    changed = True
                    break

                target_def = workflows[target]
                if not target_def.steps:
                    _log.warning(
                        "Workflow rejected: {kind} target has zero steps: "
                        "skill={skill}, workflow={workflow}, step={step}, "
                        "target_skill={ts}, target_workflow={tw}",
                        kind=edge_kind,
                        skill=vertex[0],
                        workflow=vertex[1],
                        step=step.id,
                        ts=target_skill,
                        tw=target_wf,
                    )
                    rejected.add(vertex)
                    changed = True
                    break

    # Return only the newly-rejected ones
    return rejected - already_rejected


# ---------------------------------------------------------------------------
# In-memory mutation dataclasses (used by cascade engine + repository)
# ---------------------------------------------------------------------------

@dataclass(frozen=True)
class UpdateState:
    """Queued update to an existing workflow state record."""

    layer_id: str
    step_states: dict[str, str]
    current_step: str | None
    loop_state: dict | None = None


@dataclass(frozen=True)
class CreateChild:
    """Queued creation of a child workflow state record."""

    child_id: str
    parent_id: str
    parent_step_id: str
    skill_name: str
    workflow_name: str
    step_states: dict[str, str]
    definition_snapshot: list[dict]
    scratchpad_path: str


@dataclass(frozen=True)
class SoftDelete:
    """Queued soft-delete of a workflow state record."""

    layer_id: str


@dataclass
class MutationBatch:
    """Ordered list of mutations staged in memory during cascade evaluation.

    Applied atomically by ``repository.apply_mutation_batch``.
    """

    ordered: list[UpdateState | CreateChild | SoftDelete] = field(default_factory=list)


@dataclass(frozen=True)
class CascadeOutcome:
    """Result of the cascade loop — drives response construction."""

    deepest_layer_id: str
    active_step_id: str | None  # None when top-level finalized
    condition_skips: list[tuple[str, str, str]]  # (workflow_name, step_id, reason)
    finalized_top_level: bool
    halted_at_loop_step: str | None = None  # step_id of loop step that halted auto-advance

File: src/extensions/workflows/composition.ts
import type { StepDefinition, WorkflowDefinition } from "./loader.ts";
import type { LoopState, StepSnapshot, StepStates } from "./model.ts";

// ---- vertex addressing ---------------------------------------------------------

/** A workflow is addressed by its (skill, workflow) pair, keyed as `skill/workflow`. */
export interface WorkflowRef {
  skillName: string;
  workflowName: string;
}

export const refKey = (skillName: string, workflowName: string): string =>
  `${skillName}/${workflowName}`;

/**
 * Resolve a `composes`/`loop` frontmatter value to a {@link WorkflowRef}.
 *
 * Same-skill: `"process-note"` → `{ parentSkill, "process-note" }`.
 * Cross-skill: `"other-skill/process-note"` → `{ "other-skill", "process-note" }`.
 *
 * Throws on empty or malformed input (missing half, multiple slashes).
 */
export const resolveComposes = (value: string, parentSkillName: string): WorkflowRef => {
  const trimmed = value.trim();

  if (trimmed.length === 0) throw new Error("composition value must be a non-empty string");

  if (trimmed.includes("/")) {
    const slash = trimmed.indexOf("/");
    const skill = trimmed.slice(0, slash);
    const workflow = trimmed.slice(slash + 1);

    if (skill.length === 0 || workflow.length === 0 || workflow.includes("/")) {
      throw new Error(`Malformed composition value: '${value}'`);
    }

    return { skillName: skill, workflowName: workflow };
  }

  return { skillName: parentSkillName, workflowName: trimmed };
};

/** The single composition edge a step declares (composes or loop), or null. */
export const compositionEdge = (step: StepDefinition): string | null =>
  step.composes ?? step.loop ?? null;

// ---- graph validation (mirrors the legacy registry-load checks) ----------------

/**
 * Three-color DFS that returns each cycle found in the composition graph
 * (composes and loop edges share one graph). A self-loop is a one-vertex cycle.
 */
export const detectCycles = (workflows: Map<string, WorkflowDefinition>): string[][] => {
  const adjacency = new Map<string, string[]>();

  for (const [key, def] of workflows) {
    const targets: string[] = [];

    for (const step of def.steps) {
      const edge = compositionEdge(step);

      if (edge == null) continue;

      try {
        const target = resolveComposes(edge, def.skillName);
        const targetKey = refKey(target.skillName, target.workflowName);

        if (workflows.has(targetKey)) targets.push(targetKey);
      } catch {
        // Malformed edges are caught by reference validation, not cycle detection.
      }
    }

    adjacency.set(key, targets);
  }

  const WHITE = 0;
  const GRAY = 1;
  const BLACK = 2;
  const color = new Map<string, number>([...workflows.keys()].map((key) => [key, WHITE]));
  const stack: string[] = [];
  const cycles: string[][] = [];

  const visit = (vertex: string): void => {
    color.set(vertex, GRAY);
    stack.push(vertex);

    for (const neighbour of adjacency.get(vertex) ?? []) {
      if (color.get(neighbour) === GRAY) {
        cycles.push(stack.slice(stack.indexOf(neighbour)));
      } else if (color.get(neighbour) === WHITE) {
        visit(neighbour);
      }
    }

    stack.pop();
    color.set(vertex, BLACK);
  };

  for (const vertex of workflows.keys()) {
    if (color.get(vertex) === WHITE) visit(vertex);
  }

  return cycles;
};

export interface GraphValidation {
  /** Vertex keys (`skill/workflow`) that must not run — they would cycle or dangle. */
  rejected: Set<string>;
  /** Human-readable reasons, one per rejection, for logging. */
  warnings: string[];
}

/**
 * Reject workflows whose composition is unsafe: a step declaring both `composes`
 * and `loop`, a cycle member, or a reference to a missing/empty/rejected target.
 * Rejection cascades — a parent pointing at a rejected target is itself rejected.
 */
export const validateWorkflowGraph = (
  workflows: Map<string, WorkflowDefinition>,
): GraphValidation => {
  const rejected = new Set<string>();
  const warnings: string[] = [];

  const reject = (key: string, reason: string): void => {
    if (rejected.has(key)) return;

    rejected.add(key);
    warnings.push(`${key}: ${reason}`);
  };

  // Mutex pre-pass: a step cannot be both a composition and a loop.
  for (const [key, def] of workflows) {
    for (const step of def.steps) {
      if (step.composes != null && step.loop != null) {
        reject(key, `step '${step.id}' declares both composes and loop`);
        break;
      }
    }
  }

  for (const cycle of detectCycles(workflows)) {
    for (const member of cycle) reject(member, `composition cycle: ${cycle.join(" → ")}`);
  }

  // Fixed-point: rejecting a target may cascade to its parents.
  let changed = true;

  while (changed) {
    changed = false;

    for (const [key, def] of workflows) {
      if (rejected.has(key)) continue;

      for (const step of def.steps) {
        const edge = compositionEdge(step);

        if (edge == null) continue;

        const kind = step.loop != null ? "loop" : "composes";
        let targetKey: string;

        try {
          const target = resolveComposes(edge, def.skillName);
          targetKey = refKey(target.skillName, target.workflowName);
        } catch {
          reject(key, `step '${step.id}' has a malformed ${kind} value: '${edge}'`);
          changed = true;
          break;
        }

        const target = workflows.get(targetKey);

        if (target == null) {
          reject(key, `step '${step.id}' ${kind} target '${targetKey}' is missing`);
          changed = true;
          break;
        }

        if (rejected.has(targetKey)) {
          reject(key, `step '${step.id}' ${kind} target '${targetKey}' was itself rejected`);
          changed = true;
          break;
        }

        if (target.steps.length === 0) {
          reject(key, `step '${step.id}' ${kind} target '${targetKey}' has no steps`);
          changed = true;
          break;
        }
      }
    }
  }

  return { rejected, warnings };
};

// ---- in-memory mutations staged during a cascade -------------------------------

/** Update an existing layer's step states / current step / loop bookkeeping. */
export interface UpdateStateMutation {
  kind: "update";
  layerId: string;
  stepStates: StepStates;
  currentStep: string | null;
  /** Present only when the cascade touched loop bookkeeping for this layer. */
  loopState?: LoopState;
}

/** Spawn a composed/iterated child layer. */
export interface CreateChildMutation {
  kind: "create";
  childId: string;
  parentId: string;
  parentStepId: string;
  skillName: string;
  workflowName: string;
  stepStates: StepStates;
  definitionSnapshot: StepSnapshot[];
  scratchpadPath: string;
}

/** Soft-delete a layer (a finished child, or the finalized top-level). */
export interface SoftDeleteMutation {
  kind: "softDelete";
  layerId: string;
}

export type Mutation = UpdateStateMutation | CreateChildMutation | SoftDeleteMutation;

/** Ordered mutations applied atomically by the repository. */
export type MutationBatch = Mutation[];

// ---- cascade results -----------------------------------------------------------

export interface CascadeOutcome {
  deepestLayerId: string;
  /** The step now active (started), or null when the top-level finalized. */
  activeStepId: string | null;
  finalizedTopLevel: boolean;
  /** Tally of the finalized top-level layer's own steps (set on finalize). */
  completedCount?: number;
  skippedCount?: number;
  /** Auto-advance halted at this loop step — agent must start it with items. */
  haltedAtLoopStep?: string;
  /** Auto-advance halted at this conditional step — agent decides start/skip. */
  haltedAtConditionStep?: string;
}

/** One segment of the active-path breadcrumb. */
export interface BreadcrumbPart {
  workflowName: string;
  stepId: string | null;
  item: string | null;
}
